ocaml/xapi/xapi_observer.ml

```diff
@@ -110,14 +110,128 @@ module Observer : ObserverInterface = struct
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_attributes dbg uuid attributes
+
+    let set_endpoints ~__context ~uuid ~endpoints =
+      debug "Observer.set_endpoints %s" uuid ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_endpoints dbg uuid endpoints
+
+    let init ~__context =
+      debug "Observer.init" ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.init dbg
+
+    let set_trace_log_dir ~__context ~dir =
+      debug "Observer.set_trace_log_dir" ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_trace_log_dir dbg dir
+
+    let set_export_interval ~__context ~interval =
+      debug "Observer.set_export_interval" ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_export_interval dbg interval
+
+    let set_max_spans ~__context ~spans =
+      debug "Observer.set_max_spans" ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_max_spans dbg spans
+
+    let set_max_traces ~__context ~traces =
+      debug "Observer.set_max_traces" ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_max_traces dbg traces
+
+    let set_max_file_size ~__context ~file_size =
+      debug "Observer.set_max_file_size" ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_max_file_size dbg file_size
+
+    let set_host_id ~__context ~host_id =
+      debug "Observer.set_host_id" ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_host_id dbg host_id
+
+    let set_compress_tracing_files ~__context ~enabled =
+      debug "Observer.set_compress_tracing_files" ;
+      let module S = (val local_client ~__context : XAPI_CLUSTER) in
+      let dbg = Context.string_of_task __context in
+      S.Observer.set_compress_tracing_files dbg enabled
+  end
+end
+
+let supported_components = ["xapi"; "xenopsd"; "xapi-clusterd"]
+
+(* We startup the observer for clusterd separately in cluster_host so that
+   there is no need to restart xapi in order for clusterd to be observed.
+   This does mean that observer will always be enabled for clusterd. *)
+let startup_components =
+  List.filter (( <> ) "xapi-clusterd") supported_components
 
 let get_forwarder c =
   let module Forwarder = ( val match c with
```

Comment:
I don't like that this is based on strings... Can we switch to a variant type?

Reply:
there is some complication where we need to store the components to a database, and converting them back to strings (e.g. the `register` function in `xapi_observer.ml`) . I am not sure if it's still worth adding a variant type

Reply:
We can add the to and from string function to this though, when we read from/write to the database

Reply:
done as separate PR #5269